sound/soc/intel/boards/ehl_rt5660.c

```diff
@@ -0,0 +1,387 @@
+static const char cname[] = "i2c-INT34C2:00";
+#else
+static const char pname[] = "0000:00:1f.3";
+static const char cname[] = "i2c-INTC1027:00"; /* EHL Board */
+#endif
+
+SND_SOC_DAILINK_DEF(ssp0_pin,
+	DAILINK_COMP_ARRAY(COMP_CPU("SSP0 Pin")));
+
+SND_SOC_DAILINK_DEF(rt5660_codec,
+	DAILINK_COMP_ARRAY(COMP_CODEC(cname, RT5660_CODEC_DAI)));
+
+SND_SOC_DAILINK_DEF(platform,
+	DAILINK_COMP_ARRAY(COMP_PLATFORM(pname)));
+
+SND_SOC_DAILINK_DEF(dmic_pin,
+	DAILINK_COMP_ARRAY(COMP_CPU("DMIC01 Pin")));
+SND_SOC_DAILINK_DEF(dmic_codec,
+	DAILINK_COMP_ARRAY(COMP_CODEC("dmic-codec", "dmic-hifi")));
+SND_SOC_DAILINK_DEF(dmic16k,
+	DAILINK_COMP_ARRAY(COMP_CPU("DMIC16k Pin")));
+
+#if IS_ENABLED(CONFIG_SND_SOC_HDAC_HDMI)
+SND_SOC_DAILINK_DEF(idisp1_pin,
+	DAILINK_COMP_ARRAY(COMP_CPU("iDisp1 Pin")));
+SND_SOC_DAILINK_DEF(idisp1_codec,
+	DAILINK_COMP_ARRAY(COMP_CODEC("ehdaudio0D2", "intel-hdmi-hifi1")));
+
+SND_SOC_DAILINK_DEF(idisp2_pin,
+	DAILINK_COMP_ARRAY(COMP_CPU("iDisp2 Pin")));
+SND_SOC_DAILINK_DEF(idisp2_codec,
+	DAILINK_COMP_ARRAY(COMP_CODEC("ehdaudio0D2", "intel-hdmi-hifi2")));
+
+SND_SOC_DAILINK_DEF(idisp3_pin,
+	DAILINK_COMP_ARRAY(COMP_CPU("iDisp3 Pin")));
+SND_SOC_DAILINK_DEF(idisp3_codec,
+	DAILINK_COMP_ARRAY(COMP_CODEC("ehdaudio0D2", "intel-hdmi-hifi3")));
+
+SND_SOC_DAILINK_DEF(idisp4_pin,
+	DAILINK_COMP_ARRAY(COMP_CPU("iDisp4 Pin")));
+SND_SOC_DAILINK_DEF(idisp4_codec,
+	DAILINK_COMP_ARRAY(COMP_CODEC("ehdaudio0D2", "intel-hdmi-hifi4")));
+#endif	/* IS_ENABLED(CONFIG_SND_SOC_HDAC_HDMI) */
+
+static struct snd_soc_dai_link ehl_rt5660_msic_dailink[] = {
+	/* back ends */
+	{
+		.name = "SSP0-Codec",
+		.id = 0,
+		.init = NULL,
+		.no_pcm = 1,
+		.dai_fmt = SND_SOC_DAIFMT_I2S |
+			SND_SOC_DAIFMT_NB_NF | SND_SOC_DAIFMT_CBS_CFS,
+		.dpcm_playback = 1,
+		.dpcm_capture = 1,
+		.ops = &ehl_rt5660_ops,
+		.be_hw_params_fixup = ehl_be_fixup,
+		.nonatomic = true,
+		SND_SOC_DAILINK_REG(ssp0_pin, rt5660_codec, platform),
+	},
+	{
+		.name = "dmic48k",
+		.id = 1,
+		.ignore_suspend = 1,
+		.dpcm_capture = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(dmic_pin, dmic_codec, platform),
+	},
+	{
+		.name = "dmic16k",
+		.id = 2,
+		.ignore_suspend = 1,
+		.dpcm_capture = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(dmic16k, dmic_codec, platform),
+	},
+#if IS_ENABLED(CONFIG_SND_SOC_HDAC_HDMI)
+	{
+		.name = "iDisp1",
+		.id = 5,
+		.init = ehl_hdmi_init,
+		.dpcm_playback = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(idisp1_pin, idisp1_codec, platform),
+	},
+	{
+		.name = "iDisp2",
+		.id = 6,
+		.init = ehl_hdmi_init,
+		.dpcm_playback = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(idisp2_pin, idisp2_codec, platform),
+	},
+	{
+		.name = "iDisp3",
+		.id = 7,
+		.init = ehl_hdmi_init,
+		.dpcm_playback = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(idisp3_pin, idisp3_codec, platform),
+	},
+	{
+		.name = "iDisp4",
+		.id = 8,
+		.init = ehl_hdmi_init,
+		.dpcm_playback = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(idisp4_pin, idisp4_codec, platform),
+	},
```

Comment:
can we follow the example of sof_rt5682.c and create the dai links as in the create_dai_links() method?

Reply:
@ranj063 I saw there are some other files are using SND_SOC_DAILINK_REG() like kbl_rt5660.c, skl_hda_dsp_common.c and so on. It there a rule that in which drivers we should use SND_SOC_DAILINK_REG() and in which drivers we should use sof_rt5682-liked method?

Reply:
> @ranj063 I saw there are some other files are using SND_SOC_DAILINK_REG() like kbl_rt5660.c, skl_hda_dsp_common.c and so on. It there a rule that in which drivers we should use SND_SOC_DAILINK_REG() and in which drivers we should use sof_rt5682-liked method?

Use the macro,that's the rule. I don't know why sof-rt5682 didn't use it, maybe it was in flight when this was added. @bardliao you may want to clean this up :-)

Reply:
> SND_SOC_DAILINK_REG

@plbossart To be honest, I tried to use SND_SOC_DAILINK_REG() in sof_rt5682.c, but I didn't find a good way to do so for the dai links are created dynamically.

Reply:
@plbossart Ok with the explanation?

Reply:
let's stay with fixed dailinks, this is a very specific platform and I am not aware of any other use of RT5660. We can revisit this later, it's no big deal.